homeassistant/components/habitica/button.py

```diff
@@ -0,0 +1,148 @@
+"""Habitica button platform."""
+
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+from enum import StrEnum
+from http import HTTPStatus
+from typing import TYPE_CHECKING, Any
+
+from aiohttp import ClientResponseError
+
+from homeassistant.components.button import ButtonEntity, ButtonEntityDescription
+from homeassistant.const import CONF_NAME, CONF_URL
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ServiceValidationError
+from homeassistant.helpers.device_registry import DeviceEntryType, DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from . import HabiticaConfigEntry
+from .const import DOMAIN, MANUFACTURER, NAME
+from .coordinator import HabiticaData, HabiticaDataUpdateCoordinator
+
+
+@dataclass(kw_only=True, frozen=True)
+class HabiticaButtonEntityDescription(ButtonEntityDescription):
+    """Describes Habitica button entity."""
+
+    press_fn: Callable[[HabiticaDataUpdateCoordinator], Any]
+    available_fn: Callable[[HabiticaData], bool] | None = None
+
+
+class HabitipyButtonEntity(StrEnum):
+    """Habitica button entities."""
+
+    RUN_CRON = "run_cron"
+    BUY_HEALTH_POTION = "buy_health_potion"
+    ALLOCATE_ALL_STAT_POINTS = "allocate_all_stat_points"
+    REVIVE = "revive"
+
+
+BUTTON_DESCRIPTIONS: tuple[HabiticaButtonEntityDescription, ...] = (
+    HabiticaButtonEntityDescription(
+        key=HabitipyButtonEntity.RUN_CRON,
+        translation_key=HabitipyButtonEntity.RUN_CRON,
+        press_fn=lambda coordinator: coordinator.api.cron.post(),
+        available_fn=lambda data: data.user["needsCron"],
+    ),
+    HabiticaButtonEntityDescription(
+        key=HabitipyButtonEntity.BUY_HEALTH_POTION,
+        translation_key=HabitipyButtonEntity.BUY_HEALTH_POTION,
+        press_fn=(
+            lambda coordinator: coordinator.api["user"]["buy-health-potion"].post()
+        ),
+        available_fn=(
+            lambda data: data.user["stats"]["gp"] >= 25
+            and data.user["stats"]["hp"] < 50
+        ),
```

Comment:
Is this something from habitica self or is this what we think is the moment to do this

Reply:
That are the required conditions to be able to execute this feature. The API would normally return an error message like "Not enough gold" or "Already at maximum health", but with the current library these error messages are not passed through. I can only check if there was a 401 error. With a new library this will be taken into account...

Reply:
These are the 2 possible error conditions as stated in the docs:
https://habitica.com/apidoc/#api-User-UserBuyPotion